localstack-core/localstack/services/apigateway/next_gen/execute_api/handlers/resource_router.py

```diff
@@ -0,0 +1,127 @@
+import logging
+from functools import cache
+
+from werkzeug.exceptions import MethodNotAllowed
+from werkzeug.routing import Map, MapAdapter
+
+from localstack.aws.api.apigateway import ListOfResource, Resource
+from localstack.aws.protocol.op_router import (
+    GreedyPathConverter,
+    # TODO: all under are private, not sure exactly how we should proceed for them to be re-usable
+    _path_param_regex,
+    _post_process_arg_name,
+    _StrictMethodRule,
+    _transform_path_params_to_rule_vars,
+)
+from localstack.http import Response
+from localstack.services.apigateway.models import RestApiDeployment
+
+from ..api import RestApiGatewayHandler, RestApiGatewayHandlerChain
+from ..context import RestApiInvocationContext
+from ..helpers import get_resources_from_deployment
+
+LOG = logging.getLogger(__name__)
+
+
+class RestAPIResourceRouter:
+    """
+    A router implementation which abstracts the routing of incoming REST API Context to a specific
+    resource of the Deployment.
+    """
+
+    _map: Map
+
+    def __init__(self, deployment: RestApiDeployment):
+        resources = get_resources_from_deployment(deployment)
+        self._resources_map = {resource["id"]: resource for resource in resources}
+        self._map = get_rule_map_for_resources(resources)
+
+    def match(self, context: RestApiInvocationContext) -> tuple[Resource, dict[str, str]]:
+        """
+        Matches the given request to the resource it targets (or raises an exception if no resource matches).
+
+        :param context:
+        :return: A tuple with the matched resource and the (already parsed) path params
+        :raises: TODO: Gateway exception in case the given request does not match any operation
+        """
+
+        request = context.request
+        # bind the map to get the actual matcher
+        matcher: MapAdapter = self._map.bind(context.request.host)
+
+        # perform the matching
+        try:
+            # trailing slashes are ignored in APIGW
+            path = context.invocation_request["path"].rstrip("/")
+
+            rule, args = matcher.match(path, method=request.method, return_rule=True)
+        except MethodNotAllowed as e:
+            # MethodNotAllowed (405) exception is raised if a path is matching, but the method does not.
+            # Our router handles this as a 404.
+            # TODO: raise proper Gateway exception
+            raise Exception("Not found") from e
```

Comment:
Should we also catch the exception raised if no path are matching in order to raise the proper Gateway Exception also?

Reply:
Great catch! Thanks, I'll add this right now 🙏